### PR TITLE
Feature/data 776 env dependent config

### PR DESCRIPTION
**Pull request by laural21**
https://choco.atlassian.net/browse/DATA-776

Update dagger to process environment dependent configs.

### CODE REVIEW

File: dagger/config_finder/config_processor.py [hunk 1]
```diff
@@ -29,6 +31,20 @@ def _load_yaml(yaml_path):
...
+            config["template_parameters"].update(env_dependent_params.get("template_parameters", {}))
+        except AttributeError:
+            config["template_parameters"] = env_dependent_params.get("template_parameters", {})
+        try:
+            config["airflow_task_parameters"].update(env_dependent_params.get("airflow_task_parameters", {}))
+        except AttributeError:
+            config["airflow_task_parameters"] = env_dependent_params.get("airflow_task_parameters", {})
+        deactivate = env_dependent_params.get("deactivate")
+        return config, deactivate
+
```
Comment (siklosid) [inline dagger/config_finder/config_processor.py:47]: Let's use deepmerge library, it can handle null values

File: dagger/config_finder/config_processor.py [hunk 2]
```diff
@@ -48,12 +64,14 @@ def process_pipeline_configs(self):
 
                 _logger.info("Processing task config: %s", task_config_path)
                 task_config = self._load_yaml(task_config_path)
+                task_config, deactivate = self.overwrite_params(task_config)
```
Comment (siklosid) [inline dagger/config_finder/config_processor.py:67]: What if overwrite params returns None in case of deactivate and then if task_config is None we skip it

File: dagger/pipeline/pipeline.py [hunk 3]
```diff
@@ -59,12 +65,14 @@ def __init__(self, directory: str, config: dict):
...
-        self._start_date = self.parse_attribute(attribute_name="start_date")
-        self._parameters = self.parse_attribute(attribute_name="dag_parameters") or {}
+        env_dependent_config = config.get("environments").get(ENV) if config.get("environments") else {}
+
+        self._owner = env_dependent_config.get("owner") or self.parse_attribute(attribute_name="owner")
+        self._description = env_dependent_config.get("description") or self.parse_attribute(attribute_name="description")
+        self._default_args = env_dependent_config.get("default_args") or self.parse_attribute(attribute_name="default_args") or {}
+        self._schedule = env_dependent_config.get("schedule") or self.parse_attribute(attribute_name="schedule")
+        self._start_date = env_dependent_config.get("start_date") or self.parse_attribute(attribute_name="start_date")
+        self._parameters = env_dependent_config.get("dag_parameters") or self.parse_attribute(attribute_name="dag_parameters") or {}
```
Comment (siklosid) [inline dagger/pipeline/pipeline.py:75]: Let's do the same as we are doing with tasks, preferably with the same code. Read in the dictionary, check if there is env dependent config and merge it on the default parameters.

File: dockers/airflow/Dockerfile [hunk 4]
```diff
@@ -48,10 +48,18 @@ RUN curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2
 
 # Install python libraries & cleanup
 RUN pip install -U pip setuptools wheel pytest-runner
+<<<<<<< Updated upstream
```
Comment (siklosid) [inline dockers/airflow/Dockerfile:51]: Conflict remained here

File: dockers/airflow/Dockerfile [hunk 5]
```diff
@@ -48,10 +48,18 @@ RUN curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2
...
     authlib==0.15.4 \
     requests==2.25.1 \
     apache-airflow[slack,postgres,s3,statsd]==$AIRFLOW_VERSION --constraint $CONSTRAINTS \
+RUN pip install \
+    pytz \
+    pyOpenSSL \
+    ndg-httpsclient \
+    apache-airflow[crypto,slack,druid,postgres,s3,statsd]==$AIRFLOW_VERSION \
+    psycopg2-binary \
+    -U awscli>=1.16 \
```
Comment (siklosid) [inline dockers/airflow/Dockerfile:62]: What happened here?

**Review by siklosid — CHANGES_REQUESTED** [review]

File: dagger/config_finder/config_processor.py [hunk 6]
```diff
@@ -1,18 +1,22 @@
...
 from dagger.pipeline.task_factory import TaskFactory
 
-# import dagger.conf as conf
+import dagger.conf as conf
 
 
 _logger = logging.getLogger("configFinder")
 DAG_DIR = join(environ.get("AIRFLOW_HOME", "./"), "dags")
 
+ENV = conf.ENV
```
Comment (siklosid) [inline dagger/config_finder/config_processor.py:18]: Why do we need to make it another global variable if we can use it as conf.ENV?

File: dagger/config_finder/config_processor.py [hunk 7]
```diff
@@ -40,20 +52,24 @@ def process_pipeline_configs(self):
...
                 task_name = splitext(task_config.config)[0]
                 task_config_path = join(pipeline_config.directory, task_config.config)
 
                 _logger.info("Processing task config: %s", task_config_path)
-                task_config = self._load_yaml(task_config_path)
-                task_type = task_config["type"]
-                pipeline.add_task(
-                    self._task_factory.create_task(
-                        task_type, task_name, pipeline_name, pipeline, task_config
+                task_config = EnvYAML(task_config_path).export()
```
Comment (siklosid) [inline dagger/config_finder/config_processor.py:64]: What is the reason we are not using the EnvYAML for the pipeline config, only for the task config?

File: dagger/config_finder/config_processor.py [hunk 8]
```diff
@@ -29,6 +33,14 @@ def _load_yaml(yaml_path):
                 exit(1)
         return config
 
+    @staticmethod
+    def overwrite_params(config):
```
Comment (siklosid) [inline dagger/config_finder/config_processor.py:37]: Can we give a more descriptive name? localising_params maybe or something that expresses we are modifying based on environment

File: dagger/config_finder/config_processor.py [hunk 9]
```diff
@@ -40,20 +52,24 @@ def process_pipeline_configs(self):
...
 
                 _logger.info("Processing task config: %s", task_config_path)
-                task_config = self._load_yaml(task_config_path)
-                task_type = task_config["type"]
-                pipeline.add_task(
-                    self._task_factory.create_task(
-                        task_type, task_name, pipeline_name, pipeline, task_config
+                task_config = EnvYAML(task_config_path).export()
+                task_config = self.overwrite_params(task_config)
+                if task_config:
```
Comment (siklosid) [inline dagger/config_finder/config_processor.py:66]: We might want to log that the task was deactivated in the given ENV

File: dagger/pipeline/pipeline.py [hunk 10]
```diff
@@ -27,6 +27,10 @@ def init_attributes(cls, orig_cls):
                     validator=lambda x: datetime.strptime(x, "%Y-%m-%dT%H:%M"),
                 ),
                 Attribute(attribute_name="airflow_parameters"),
+                Attribute(
+                    attribute_name="environments",
+                    required=False,
+                ),
```
Comment (siklosid) [inline dagger/pipeline/pipeline.py:33]: Do we still need this attribute?

File: dagger/pipeline/task.py [hunk 11]
```diff
@@ -47,6 +47,12 @@ def init_attributes(cls, orig_cls):
                     format_help="dictionary",
                 ),
                 Attribute(attribute_name="task_parameters", nullable=True),
+                Attribute(
+                    attribute_name="environments",
+                    required=False,
+                    nullable=True,
+                    format_help="dictionary",
+                ),
```
Comment (siklosid) [inline dagger/pipeline/task.py:55]: And this?

File: dagger/pipeline/task.py [hunk 12]
```diff
@@ -63,6 +69,7 @@ def __init__(self, name: str, pipeline_name, pipeline, config: dict):
         self._airflow_parameters = self.parse_attribute("airflow_task_parameters") or {}
         self._render_parameters(self._airflow_parameters)
         self._template_parameters = self.parse_attribute("template_parameters") or {}
+        self._environments = self.parse_attribute("environments") or {}
```
Comment (siklosid) [inline dagger/pipeline/task.py:72]: And this?

File: dagger/pipeline/task.py [hunk 13]
```diff
@@ -119,6 +126,10 @@ def template_parameters(self) -> dict:
     def inputs(self) -> List[IO]:
         return self._inputs
 
+    @property
+    def environments(self) -> dict:
+        return self._environments
```
Comment (siklosid) [inline dagger/pipeline/task.py:131]: End this?

File: dockers/airflow/Dockerfile [hunk 14]
```diff
@@ -81,4 +81,4 @@ COPY docker-entrypoint.sh $AIRFLOW_HOME/docker-entrypoint.sh
 COPY airflow.cfg $AIRFLOW_HOME/airflow.cfg
 
 WORKDIR $AIRFLOW_HOME
-ENTRYPOINT ["./docker-entrypoint.sh"]
+ENTRYPOINT ["./docker-entrypoint.sh"]
```
Comment (siklosid) [inline dockers/airflow/Dockerfile:84]: ?

File: reqs/base.txt [hunk 15]
```diff
@@ -1,3 +1,4 @@
+croniter==0.3.29
```
Comment (siklosid) [inline reqs/base.txt:1]: How is this relevant for this change?

File: reqs/test.txt [hunk 16]
```diff
@@ -1,2 +1,3 @@
 pytest==6.2.5
 pytest-cov
+apache-airflow[postgres,s3,statsd]==2.2.2
```
Comment (siklosid) [inline reqs/test.txt:3]: How is this relevant for this change?

File: tests/fixtures/config_finder/root/dags/test_spark/spark_disable.yaml [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+task_parameters:
+  job_file: 'abc/test.py'
+  spark_engine: emr
+  job_args:
+      param1: test
+pool: emr
+
+environments:
+  local:
+    deactivate: True
```
Comment (siklosid) [inline tests/fixtures/config_finder/root/dags/test_spark/spark_disable.yaml:34]: New line?

**Review by siklosid — CHANGES_REQUESTED** [review]

File: dockers/airflow/Dockerfile (hunk 14 above)
Comment (laural21) [reply dockers/airflow/Dockerfile:84]: Not sure how this happened

File: reqs/base.txt (hunk 15 above)
Comment (laural21) [reply reqs/base.txt:1]: Tests don't run without installing this library

File: reqs/test.txt (hunk 16 above)
Comment (laural21) [reply reqs/test.txt:3]: Tests don't run without installing this library. It is also installed in reqs/dev.txt but we don't need all the other packages so I just included it test.txt as well

File: dagger/config_finder/config_processor.py [hunk 18]
```diff
@@ -52,8 +50,8 @@ def process_pipeline_configs(self):
             config_path = join(pipeline_config.directory, pipeline_config.config)
 
             _logger.info("Processing config: %s", config_path)
-            config_dict = self._load_yaml(config_path)
-            config_dict = self.overwrite_params(config_dict)
+            config_dict = EnvYAML(config_path).export()
+            config_dict = self.localize_params(config_dict)
```
Comment (siklosid) [inline dagger/config_finder/config_processor.py:54]: Just a small thing, now we have a `_load_yaml` function that is not used anywhere and we have this two lines replicated which is reading the config with EnvYAML and localising the params. I would put these two lines in the `_load_yaml` function, so we don't have duplication. Wdyt?

File: reqs/test.txt (hunk 16 above)
Comment (siklosid) [reply reqs/test.txt:3]: That part I understand, but I think it was added even before in a different PR. Otherwise test wouldn't have passed. So I'm just not sure why it is added here.

Comment (siklosid) [reply reqs/test.txt:3]: Or I don't understand how the tests were passing before.

Comment (laural21) [reply reqs/test.txt:3]: Should I remove it from the pr?

File: dagger/config_finder/config_processor.py [hunk 19]
```diff
@@ -21,18 +21,12 @@ def __init__(self, config_finder: ConfigFinder):
...
-            except yaml.YAMLError as exc:
-                _logger.exception("Couldn't read config file {}", yaml_path, exc)
-                exit(1)
-        return config
+    def _load_yaml(self, yaml_path):
+        config_dict = EnvYAML(yaml_path).export()
+        config_dict = self.localize_params(config_dict)
+        return config_dict
 
-    @staticmethod
```
Comment (siklosid) [inline dagger/config_finder/config_processor.py]: I don't see where you are using self in this function so not sure why it became non static

**Review by siklosid — APPROVED** [review]
